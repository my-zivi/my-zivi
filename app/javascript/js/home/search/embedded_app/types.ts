/* eslint-disable camelcase */

export interface JobPostingSearchHit {
  objectID: string;
  title: string;
  icon_url: string;
  publication_date: string;
  plain_description: string;
  description: string;
  brief_description: string;
  required_skills?: string;
  preferred_skills?: string;
  canton: string;
  identification_number: string;
  relevancy: Relevancy;
  contact_information: string;
  minimum_service_months: number;
  language: string;
  organization_display_name: string;
  category_display_name: string;
  sub_category_display_name?: string;
  link: string;
  featured_as_new: boolean;
  priority_program: boolean;
}

export enum Relevancy {
  Normal = 1,
  Urgent,
  Featured
}
