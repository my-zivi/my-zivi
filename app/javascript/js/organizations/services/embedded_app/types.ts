export interface Service {
  beginning: string;
  ending: string;
  civilServant: {
    fullName: string;
  };
}
