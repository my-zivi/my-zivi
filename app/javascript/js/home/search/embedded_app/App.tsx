import algoliasearch from 'algoliasearch';
import React from 'preact/compat';
import { createBrowserLocalStorageCache } from '@algolia/cache-browser-local-storage';
import { createFallbackableCache } from '@algolia/cache-common';
import { createInMemoryCache } from '@algolia/cache-in-memory';
import aa from 'search-insights';
import Turbo from '@hotwired/turbo-rails';
import SearchPage from '~/js/home/search/embedded_app/components/SearchPage';

declare global {
  interface Window {
    Turbo: typeof Turbo;
  }
}

aa('init', {
  appId: MyZivi.algolia.applicationId,
  apiKey: MyZivi.algolia.apiKey,
  useCookie: true,
});

const searchClient = algoliasearch(
  MyZivi.algolia.applicationId,
  MyZivi.algolia.apiKey,
  {
    responsesCache: createFallbackableCache({
      caches: [
        createBrowserLocalStorageCache({ key: 'my-zivi-search', localStorage: window.sessionStorage }),
        createInMemoryCache(),
      ],
    }),
    requestsCache: createInMemoryCache({ serializable: false }),
  },
);

const App: React.FunctionComponent = () => (
  <SearchPage searchClient={searchClient} />
);

export default App;
