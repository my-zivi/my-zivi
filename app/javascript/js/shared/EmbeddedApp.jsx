import React, { render, h } from 'preact';

export default class EmbeddedApp {
  constructor(selector, RootComponent) {
    this.selector = selector;
    this.RootComponent = RootComponent;
    this.installed = false;
  }

  install() {
    this.attachToContainer();
    $(document).on('turbolinks:load', () => this.attachToContainer());
  }

  attachToContainer() {
    const container = document.querySelector(this.selector);

    if (container && !this.installed) {
      const { RootComponent } = this;

      render(<RootComponent />, container);
      this.installed = true;
    }
  }
}
