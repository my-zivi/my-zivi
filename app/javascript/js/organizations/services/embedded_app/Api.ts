import { Service } from 'js/organizations/services/embedded_app/types';
import Rails from '@rails/ujs';

export default class Api {
  static async fetchData() {
    return new Promise((success: (services: Service[]) => void, error) => {
      Rails.ajax({
        url: '/organizations/services.json',
        type: 'GET',
        success,
        error,
      });
    });
  }
}
