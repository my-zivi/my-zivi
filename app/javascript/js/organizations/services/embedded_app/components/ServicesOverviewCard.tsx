import { Component, ComponentChild } from 'preact';
import { DATE_FORMATS } from 'js/constants';
import React from 'preact/compat';
import { Service } from 'js/organizations/services/embedded_app/types';
import Spinner from 'js/shared/components/Spinner';
import withApi, { WithApiProps } from 'js/shared/withApiHelper';
import ServicesList from '../models/ServicesList';
import OverviewTable from './OverviewTable';
import CardHeader from './CardHeader';

interface State {
  loading: boolean;
  services: Array<Service>;
  error: string | undefined;
}

export class ServicesOverviewCardImpl extends Component<WithApiProps, State> {
  constructor() {
    super();

    this.state = {
      loading: true,
      services: [],
      error: null,
    };
  }

  get servicesList(): ServicesList {
    return new ServicesList(this.state.services);
  }

  async componentDidMount(): Promise<void> {
    await this.reloadPlanningView();
  }

  async reloadPlanningView(): Promise<void> {
    try {
      const services = await this.props.api.fetchServices();

      this.setState({
        loading: false,
        services,
      });
    } catch (e) {
      let cause = MyZivi.translations.servicesOverview.errors.cannotLoad;
      if (typeof e === 'object' && 'error' in e) {
        cause = `${cause} (${e.error})`;
      }

      this.setState({
        loading: false,
        error: cause,
      });
    }
  }

  private static getOverviewSubtitle(servicesList: ServicesList): string {
    const formattedBeginning = servicesList.planBeginning.format(DATE_FORMATS.short);
    const formattedEnding = servicesList.planEnding.format(DATE_FORMATS.short);

    return `${formattedBeginning} bis ${formattedEnding}`;
  }

  render(): ComponentChild {
    if (this.state.loading) {
      return (
        <div className="card mb-3 mb-lg-0">
          <CardHeader overviewSubtitle={MyZivi.translations.loading} />
          <div className="card-body">
            <Spinner style={'info'} />
          </div>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div className="card mb-3 mb-lg-0">
          <CardHeader overviewSubtitle={MyZivi.translations.error} />
          <div className="card-body">
            {this.state.error}
          </div>
        </div>
      );
    }

    const { servicesList } = this;

    return (
      <div className="card mb-3 mb-lg-0">
        <CardHeader overviewSubtitle={ServicesOverviewCardImpl.getOverviewSubtitle(servicesList)} />

        <div className="card-body services-overview-container">
          <OverviewTable servicesList={servicesList} />
        </div>
      </div>
    );
  }
}

export default withApi(ServicesOverviewCardImpl);
