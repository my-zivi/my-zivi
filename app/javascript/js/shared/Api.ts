import Rails from '@rails/ujs';
import { Service } from '~/js/organizations/services/embedded_app/types';

interface RequestOptions {
  url: string;
  type: 'GET' | 'POST';

  [name: string]: unknown;
}

export interface IApi {
  fetchServices(): Promise<Array<Service>>;
}

export default class Api implements IApi {
  async fetchServices(): Promise<Array<Service>> {
    return this.request<Array<Service>>({
      url: MyZivi.paths.servicesOverview,
      type: 'GET',
    });
  }

  // eslint-disable-next-line class-methods-use-this
  private request<T>(options: RequestOptions): Promise<T> {
    return new Promise((success: (response: T) => void, error: (cause: unknown) => void) => {
      Rails.ajax({ ...options, success, error });
    });
  }
}
