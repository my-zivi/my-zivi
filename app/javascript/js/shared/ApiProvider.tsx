import React from 'preact/compat';
import { ComponentChild } from 'preact';
import Api, { IApi } from '~/js/shared/Api';

export const ApiContext = React.createContext<IApi>(undefined);

export default class ApiProvider extends React.Component {
  private api = new Api();

  render(): ComponentChild {
    return (
      <ApiContext.Provider value={this.api}>
        {this.props.children}
      </ApiContext.Provider>
    );
  }
}
