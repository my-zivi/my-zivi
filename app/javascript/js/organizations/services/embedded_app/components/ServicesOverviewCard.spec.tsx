import { shallow } from 'enzyme';
import React from 'preact/compat';
import Rails from 'js/shared/__mocks__/Rails';
import Factories from 'js/tests/factories/Factories';
import ServiceFactory from 'js/tests/factories/ServiceFactory';
import { ServicesOverviewCardImpl } from './ServicesOverviewCard';
import { IApi } from 'js/shared/Api';
import { Service } from 'js/organizations/services/embedded_app/types';
import { ComponentChild, ComponentType } from 'preact';
import { ApiContext } from 'js/shared/ApiProvider';

function shallowUntil(componentInstance: any, TargetComponent: ComponentType<any>, { maxTries = 10 } = {}) {
  let root = shallow(componentInstance);

  if (typeof root.type() === 'string') {
    throw new Error('Cannot unwrap this component because it is not wrapped');
  }

  do {
    root = root.dive();
  } while (!root.is(TargetComponent) && --maxTries >= 0);

  return root.shallow();
}

const fakeApi: IApi = {
  fetchServices(): Promise<Array<Service>> {
    return Promise.resolve(Factories.buildList(ServiceFactory, 2));
  },
};

function fakeWrapInApiProvider(api: IApi) {
  return (wrapped: Element, component: ComponentType<any>) => {
    return shallowUntil(
      (
        <ApiContext.Provider value={api}>
          {wrapped}
        </ApiContext.Provider>
      ),
      component,
    );
  };
}

describe('ServicesOverviewCard', () => {
  const wrapper = fakeWrapInApiProvider(fakeApi)(<ServicesOverviewCardImpl />, ServicesOverviewCardImpl);

  beforeEach(() => {
    Rails.mockResponse(Factories.buildList(ServiceFactory, 2), MyZivi.paths.servicesOverview);
  });

  it('runs', (done) => {
    setImmediate(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
