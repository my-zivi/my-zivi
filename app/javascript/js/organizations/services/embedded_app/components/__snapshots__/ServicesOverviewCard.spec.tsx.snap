// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicesOverviewCard runs 1`] = `<Consumer />`;
